framework/src/media/demux/mpeg2ts/TSParser.cpp

```diff
@@ -0,0 +1,654 @@
+							.setThreshold(2048)
+							.build();
+	if (!streamBuffer) {
+		meddbg("streamBuffer is nullptr!\n");
+		return false;
+	}
+
+	mStreamBuffer = streamBuffer;
+	mBufferReader = std::make_shared<StreamBufferReader>(mStreamBuffer);
+	mBufferWriter = std::make_shared<StreamBufferWriter>(mStreamBuffer);
+	mReaderOffset = 0;
+
+	if (!mBufferReader || !mBufferWriter) {
+		meddbg("mBufferReader/Writer is nullptr!\n");
+		return false;
+	}
+
+	mParserManager = std::make_shared<ParserManager>();
+	if (!mParserManager) {
+		meddbg("mParserManager is nullptr!\n");
+		return false;
+	}
+
+	mPESParser = std::make_shared<PESParser>();
+	if (!mPESParser) {
+		meddbg("mPESParser is nullptr!\n");
+		return false;
+	}
+
+	m_data = new uint8_t[TS_PACKET_SIZE];
+	if (m_data == nullptr) {
+		meddbg("TS packet buffer allocate failed!\n");
+		return false;
+	}
+
+	return true;
+}
+
+std::shared_ptr<ParserManager> TSParser::getParserManager(void)
+{
+	return mParserManager;
+}
+
+size_t TSParser::sizeOfSpace(void)
+{
+	return mBufferWriter->sizeOfSpace();
+}
+
+size_t TSParser::pushData(uint8_t *buf, size_t size)
+{
+	size_t written = 0;
+	if (mBufferWriter) {
+		written += mBufferWriter->write(buf, size);
+		assert(written == size);
+	}
+	return written;
+}
+
+size_t TSParser::pullData(uint8_t *buf, size_t size, prog_num_t progNum)
+{
+	if (mPESPid == -1) {
+		// setup PES pid
+		if (progNum == 0) {
+			// use default 1st program
+			std::vector<prog_num_t> programs;
+			mParserManager->GetPrograms(programs);
+			progNum = programs[0];
+		}
+		uint8_t streamType;
+		if (!mParserManager->GetAudioStreamInfo(progNum, streamType, mPESPid)) {
+			meddbg("get audio pes pid failed\n");
+			return 0;
+		}
+
+		medvdbg("setup audio pes pid: 0x%x\n", mPESPid);
+	}
+
+	size_t fill = 0;
+	size_t need;
+	while (fill < size) {
+		need = size - fill;
+		if (mPESPacket) {
+			// get remaining payload in last PES packet
+			if (need > mPESParser->ESDataLength() - mPESDataUsed) {
+				need = mPESParser->ESDataLength() - mPESDataUsed;
+			}
+
+			memcpy(&buf[fill], mPESParser->ESData() + mPESDataUsed, need);
+			mPESDataUsed += need;
+			fill += need;
+
+			if (mPESDataUsed == mPESParser->ESDataLength()) {
+				// read whole es data in one pes packet
+				delete mPESPacket;
+				mPESPacket = NULL;
+				mPESDataUsed = 0;
+			}
+			continue;
+		}
+
+		// get new PES packet
+		if (getPESPacket(&mPESPacket) <= 0) {
+			// 0 / -1
+			break;
+		}
+
+		// parse PES packet
+		if (mPESParser->Parse(mPESPacket->Data(), mPESPacket->DataLength())) {
+			mPESDataUsed = 0;
+		} else {
+			meddbg("PES parse failed\n");
+			assert(0);
+		}
+	}
+
+	return fill;
+}
+
+bool TSParser::getPrograms(std::vector<prog_num_t> &progs)
+{
+	return mParserManager->GetPrograms(progs);
+}
+
+audio_type_t TSParser::getAudioType(prog_num_t progNum)
+{
+	uint8_t streamType;
+	ts_pid_t streamPid;
+	if (mParserManager->GetAudioStreamInfo(progNum, streamType, streamPid)) {
+		switch (streamType) {
```

Comment:
If streamPid is set in getAudioStreamInfo, but it is not valid here, we can keep current work?

Reply:
streamType and streamPid is used to receive.
getAudioStreamInfo(IN progNum, OUT streamType, OUT streamPid);